src/serde.rs

```diff
@@ -0,0 +1,92 @@
+//! CID Serde (de)serialization for the IPLD Data Model.
+//!
+//! CIDs cannot directly be represented in any of the native Serde Data model types. In order to
+//! work around that limitation. a newtype struct is introduced, that is used as a marker for Serde
+//! (de)serialization.
+use std::convert::TryFrom;
+use std::fmt;
+
+use serde::{de, ser};
+
+use crate::Cid;
+
+/// The newtype struct name that is used by Serde internally.
+pub const CID_SERDE_NEWTYPE_STRUCT_NAME: &str = "$__serde__newtype__struct__for__cid";
+
+impl ser::Serialize for Cid {
+    fn serialize<S>(&self, s: S) -> Result<S::Ok, S::Error>
+    where
+        S: ser::Serializer,
+    {
+        let value = serde_bytes::ByteBuf::from(self.to_bytes());
+        s.serialize_newtype_struct(CID_SERDE_NEWTYPE_STRUCT_NAME, &value)
+    }
+}
+
+/// Visitor to deserialize a CID that is wrapped in a new type struct named as defined at
+/// [`CID_SERDE_NEWTYPE_STRUCT_NAME`].
+pub struct CidVisitor;
+
+impl<'de> de::Visitor<'de> for CidVisitor {
+    type Value = Cid;
+
+    fn expecting(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        write!(fmt, "a valid CID in bytes, wrapped in a new struct")
+    }
+
+    /// Define `visit_newtype_struct` so that we have an entry-point from the seserializer to pass
+    /// in a custom deserializer just for CIDs.
+    fn visit_newtype_struct<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        deserializer.deserialize_bytes(self)
+    }
+
+    /// Some Serde data formats interpret a byte stream as a sequence of bytes (e.g. `serde_json`).
+    fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+    where
+        A: de::SeqAccess<'de>,
+    {
+        let mut bytes = Vec::new();
+        while let Some(byte) = seq.next_element()? {
+            bytes.push(byte);
+        }
+        Cid::try_from(bytes)
+            .map_err(|err| de::Error::custom(format!("Failed to deserialize CID: {}", err)))
+    }
+
+    fn visit_bytes<E>(self, value: &[u8]) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        println!("vmx: visit bytes: value: {:?}", value);
+        Cid::try_from(value)
+            .map_err(|err| de::Error::custom(format!("Failed to deserialize CID: {}", err)))
+    }
+}
```

Comment:
I guess this lets us encode/decode when the underlying serializer/deserializer doesn't support CIDs?

Reply:
Yes, the test case below is using a stock `serde_json` and can encode/decode CIDs.

I think it deserves a longer explanation how this is supposed to work. The CID encoding is done within rust-cid. It just encodes it into bytes. It's not IPLD Codec specific. But now we can create specific codecs for IPLD (which I did for CBOR), that natively understand CIDs. There you would use this serializer to get the bytes, but you might do additional work, in case of DAG-CBOR, you would wrap it in tag 42 and prefix the CID with a null byte.

This way a potential DAG-JSON would work the same. It could also rely on this serialization, but would then take the additional step of base encoding it.

Reply:
Yeah, my question is: Should we just refuse to decode CIDs in that case?

I think the answer is actually _yes_. Otherwise, even the dag-cbor decoder will happily decode a CID from a byte array.

Reply:
Good point. I thought I've worked around that, but thinking more about that, I might not have. I should write a test for this.

Reply:
Good news, dag-cbor won't encode all binaries. I've added test here: https://github.com/vmx/cbor/commit/b2f1afcb7ea2293e9783428a56cb5deed9468b83

The reason is that I switch to a different (de)serializer when we encounter tag 42, otherwise it's the default CBOR implementation.

Reply:
Have we tried decoding directly into a `Cid`?

Reply:
Then it will surely try to parse a CID, but isn't that the whole point?

Reply:
Basically, I'm asking if an arbitrary byte slice will decode to a CID. If I have the struct:


```rust
struct Foo {
    link: Cid,
}
```

`link` should only decode if it actually sees a dag-cbor CID.

Reply:
Bytes of a CID should only decode if they are wrapped in tag 42.

A test case would be:

```rust
let cbor_cid = [
    0xd8, 0x2a, 0x58, 0x25, 0x00, 0x01, 0x55, 0x12, 0x20, 0x2c, 0x26, 0xb4, 0x6b, 0x68, 0xff,
    0xc6, 0x8f, 0xf9, 0x9b, 0x45, 0x3c, 0x1d, 0x30, 0x41, 0x34, 0x13, 0x42, 0x2d, 0x70, 0x64,
    0x83, 0xbf, 0xa0, 0xf9, 0x8a, 0x5e, 0x88, 0x62, 0x66, 0xe7, 0xae,
];
let decoded_cbor_cid: Cid = from_slice(&cbor_cid).unwrap();
assert_eq!(decoded_cbor_cid.to_bytes(), &cbor_cid[5..]);

// The CID without the tag 42 prefix
let cbor_bytes = &cbor_cid[2..];
println!("vmx: cid wiouth tag: {:02X?}", cbor_bytes);

// This should fail
let decoded_cbor_bytes: Cid = from_slice(&cbor_cid).unwrap();
println!("vmx: cid: {:?}", decoded_cbor_bytes);
```